### PR TITLE
[WIP] Extended builds

### PR DESCRIPTION
**Pull request by PI-Victor**
Trello: https://trello.com/c/MWUAjl6k/237-13-extended-builds-separate-mvn-vs-jboss-evg
This is dependent on: https://github.com/openshift/source-to-image/pull/505/files
Limited functionality.
to make this work, in your bc

```
   strategy:
      sourceStrategy:
        env:
        - name: EXAMPLE
          value: sample-app
        from:
          kind: ImageStreamTag
          name: ruby-22-centos7:latest
        runtimeArtifacts:
        - sourcePath: path/to/1
          destinationDir: path/to/1
       - sourcePath: path/to/2
          destinationDir: path/to/2
        - sourcePath: path/to/3
          destinationDir: path/to/3
        - sourcePath: path/to/4
          destinationDir: path/to/4
        runtimeImage:
          name: centos:latest
```
cc: @php-coder @bparees 

LE: changed paths in mapping to be relative

### CODE REVIEW

File: pkg/build/builder/sti.go [hunk 1]
```diff
@@ -384,3 +393,13 @@ func scriptProxyConfig(build *api.Build) (*s2iapi.ProxyConfig, error) {
 	}
 	return config, nil
 }
+
+func copyArtifactSourceList(artifactsMapping []api.ImageSourcePath) (volumeList []s2iapi.VolumeSpec) {
```
Comment (php-coder) [inline pkg/build/builder/sti.go]: Maybe name it like `asVolumeList()`? Or `toVolumeList()`? Or also use verb like `transform`? Because for me we're not copying here, but creating a new object.

File: pkg/build/builder/sti.go [hunk 2]
```diff
@@ -354,8 +362,9 @@ func buildEnvVars(build *api.Build) s2iapi.EnvironmentList {
 // the strategy's environment. There is no preference given to either lowercase
 // or uppercase form of the variable.
 func scriptProxyConfig(build *api.Build) (*s2iapi.ProxyConfig, error) {
-	httpProxy := ""
-	httpsProxy := ""
+	var httpProxy string
+	var httpsProxy string
```
Comment (php-coder) [inline pkg/build/builder/sti.go]: Why did you need to modify this code, btw?

File: pkg/build/builder/sti.go (hunk 1 above)
Comment (php-coder) [inline pkg/build/builder/sti.go]: Do we really need to use named variable for returning here? I don't know what is the code style but I personally trying to not use them.

File: pkg/build/builder/sti.go [hunk 3]
```diff
@@ -92,8 +92,14 @@ func (s *S2IBuilder) Build() error {
...
 	}
-
+	var runtimeImageName string
 	var push bool
+	runtimeImage := s.build.Spec.Strategy.SourceStrategy.RuntimeImage
+	runtimeArtifacts := copyArtifactSourceList(s.build.Spec.Strategy.SourceStrategy.RuntimeArtifacts)
+
+	if runtimeImage != nil && len(runtimeImage.Name) > 0 {
+		runtimeImageName = runtimeImage.Name
+	}
```
Comment (php-coder) [inline pkg/build/builder/sti.go]: Maybe we need to add validation on the OpenShift side. I have validation on s2i side but I'm not sure how it will work with OpenShift and also it's better to fail fast.

Comment (PI-Victor) [inline pkg/build/builder/sti.go]: didn't push the latest changes. i think this should be a part of the buildconfig validation than it should be here.

File: pkg/build/builder/sti.go (hunk 2 above)
Comment (PI-Victor) [inline pkg/build/builder/sti.go]: because i loath this way of misusing shorthand assignment for string variables.
when you define a new int do you do newInt := 0 ? do you do falseVar := false ?

Comment (php-coder) [inline pkg/build/builder/sti.go]: Ok, I don't against such changes. I can read any of the ways of defining variables. Just wanted to be sure that this is an intentional change and not the forgotten part after some refactoring.

File: pkg/build/builder/sti.go (hunk 1 above)
Comment (PI-Victor) [inline pkg/build/builder/sti.go]: 1. good point
2. that is your preference - i'd have to add another line where i declare the variable 
my preference is to code as little as i can. however -  i do agree, it shouldn't be used in short functions just like this one.
https://tour.golang.org/basics/7

File: pkg/build/builder/sti.go (hunk 2 above)
Comment (bparees) [inline pkg/build/builder/sti.go]: i don't have a strong opinion either way (though i am somewhat sympathetic to @PI-Victor's position), which perhaps motivates the next thing i'm going to say, but: creating code churn due to personal preference is not ideal.  You've changed this, and now when someone else is in this file, they decide they prefer httpProxy:="" and back it goes, etc...   

I don't see a compelling reason for us to have a "standard" on which way this particular type of declaration should be done so it seems we should default to leaving it as it was written by the original author.

Comment (PI-Victor) [inline pkg/build/builder/sti.go]: ok, will revert, no reason to spend energy over this.

File: pkg/build/builder/sti.go (hunk 1 above)
Comment (php-coder) [inline pkg/build/builder/sti.go]: If you decided to use it, could I ask you to use empty `return` then and emphasize the fact of using named return value (like in the example that you provided)?

File: pkg/build/api/types.go [hunk 4]
```diff
@@ -509,6 +509,13 @@ type SourceBuildStrategy struct {
 
 	// ForcePull describes if the builder should pull the images from registry prior to building.
 	ForcePull bool
+
+	// RuntimeImage is the optinal image that is used to run an application
```
Comment (bparees) [inline pkg/build/api/types.go]: optional

File: pkg/build/api/types.go [hunk 5]
```diff
@@ -509,6 +509,13 @@ type SourceBuildStrategy struct {
...
+	// without unneeded dependencies installed.
```
Comment (bparees) [inline pkg/build/api/types.go]: without build-time dependencies installed.

File: pkg/build/api/types.go [hunk 6]
```diff
@@ -509,6 +509,13 @@ type SourceBuildStrategy struct {
...
+	RuntimeImage *kapi.ObjectReference
+
+	// RuntimeArtifacts holds the mapping between the artifacts paths on the builder image and the paths on the runtime image where they should be copied.
```
Comment (bparees) [inline pkg/build/api/types.go]: s/on the builder/in the builder/
s/on the runtime/in the runtime/

File: pkg/build/builder/sti.go [hunk 7]
```diff
@@ -92,9 +92,16 @@ func (s *S2IBuilder) Build() error {
...
 		return fmt.Errorf("the source to image builder must be used with the source strategy")
 	}
-
+	var runtimeImageName string
 	var push bool
 
+	runtimeImage := s.build.Spec.Strategy.SourceStrategy.RuntimeImage
+	runtimeArtifacts := copyArtifactSourceList(s.build.Spec.Strategy.SourceStrategy.RuntimeArtifacts)
+
+	if runtimeImage != nil && runtimeImage.Name != "" {
```
Comment (bparees) [inline pkg/build/builder/sti.go]: if runtimeImage.Name is nil that should be a validation error.

File: pkg/build/builder/sti.go [hunk 8]
```diff
@@ -92,9 +92,16 @@ func (s *S2IBuilder) Build() error {
 	if s.build.Spec.Strategy.SourceStrategy == nil {
 		return fmt.Errorf("the source to image builder must be used with the source strategy")
 	}
-
+	var runtimeImageName string
 	var push bool
 
+	runtimeImage := s.build.Spec.Strategy.SourceStrategy.RuntimeImage
```
Comment (bparees) [inline pkg/build/builder/sti.go]: at some point you need to resolve the RuntimeImage reference to a docker pull spec, because the ObjectReference could be to an ImageStreamTag (name+namespace point to an imagestreamtag from which you can get a pullspec), or a DockerImage (name is the docker pull spec to use).

See what we do with buildImage, this will be the same:
https://github.com/openshift/origin/blob/master/pkg/build/generator/generator.go#L482-L504

File: pkg/build/api/v1/types.go [hunk 9]
```diff
@@ -477,6 +477,13 @@ type SourceBuildStrategy struct {
 
 	// forcePull describes if the builder should pull the images from registry prior to building.
 	ForcePull bool `json:"forcePull,omitempty"`
+
+	// RuntimeImage is the optinal image that is used to run an application
+	// without unneeded dependencies installed.
+	RuntimeImage *kapi.ObjectReference `json:"runtimeImage,omitempty"`
+
+	// RuntimeArtifacts holds the mapping between the artifacts paths on the builder image and the paths on the runtime image where they should be copied.
+	RuntimeArtifacts []ImageSourcePath `json:"runtimeArtifacts,omitempty"`
```
Comment (bparees) [inline pkg/build/api/v1/types.go]: we should add a separate RuntimePullSecret field.

**Comment by bparees** [comment]
i'm not going to add the label now since this is very much a work in progress, but ultimately this will need the api-review label before merging.

File: pkg/build/api/v1/types.go (hunk 9 above)
Comment (php-coder) [inline pkg/build/api/v1/types.go]: >we should add a separate RuntimePullSecret field.

I should add the same field to the s2i, right?

File: pkg/build/api/types.go (hunk 6 above)
Comment (PI-Victor) [inline pkg/build/api/types.go]: i'm not particularly satisfied with the docs, it was more to satisfy update-generated-swagger-desc, so i might rephrase this completely.

File: pkg/build/api/v1/types.go (hunk 9 above)
Comment (bparees) [inline pkg/build/api/v1/types.go]: > I should add the same field to the s2i, right?

no, you don't provide pull secrets when using s2i, you are expected to have done a docker login such that your docker config.json contains any auth that will be needed to access necessary images.

Comment (bparees) [inline pkg/build/api/v1/types.go]: unless you mean add it to the config object that's getting passed from openshift, then yes.  but it won't be exposed as a user option.

File: pkg/build/api/types.go (hunk 6 above)
Comment (bparees) [inline pkg/build/api/types.go]: :+1:

File: pkg/build/api/v1/types.go (hunk 9 above)
Comment (php-coder) [inline pkg/build/api/v1/types.go]: Yes, I was talking about adding another field to `Config`. Right now I'm re-using the same `Docker` instance that we're using for downloading builder image (with the same pull secrets).

As far I understand OpenShift doesn't download images in this case, right?

Comment (bparees) [inline pkg/build/api/v1/types.go]: > Right now I'm re-using the same Docker instance that we're using for downloading builder image (with the same pull secrets).

yeah i suspect (unfortunately) there's going to be someone who wants a different pull secret for their builder image vs their runtime image and i'd rather make that an option from day one, rather than have the single pull secret represent both values, and then have to make an api change to add a second field later (and some sort of conversion that will copy the single field value into both fields, for legacy users who only specified the one field).

> As far I understand OpenShift doesn't download images in this case, right?

not sure what you mean.  doesn't download what images in what case?

Comment (php-coder) [inline pkg/build/api/v1/types.go]: >> As far I understand OpenShift doesn't download images in this case, right?
>
>not sure what you mean. doesn't download what images in what case?

Doesn't download builder image on the node where the build gets executed.

Comment (bparees) [inline pkg/build/api/v1/types.go]: > Doesn't download builder image on the node where the build gets executed.

well the s2i code that's being executed by openshift does pull the builder image down if either:
1) the pull policy is pull always/forcepull
2) the image is not present on the node

at which point it uses the pullsecret on the buildconfig, if one is defined.  if one is not defined, it uses the service account token (which will work for pulling from the openshift cluster registry, but obviously not dockerhub/etc)

File: pkg/build/api/types.go [hunk 10]
```diff
@@ -507,8 +508,23 @@ type SourceBuildStrategy struct {
...
+
+	// RuntimeArtifacts holds the mapping between the artifacts paths in the
+	// builder image and the paths in the runtime image where they should be
+	// copied.
+	RuntimeArtifacts []ImageSourcePath
+
+	// RuntimePullSecret is the name of a Secret that would be used for setting up
+	// the authentication for pulling the Docker runtime images from private
+	// Docker registries.
+	RuntimePullSecret *kapi.LocalObjectReference
```
Comment (liggitt) [inline pkg/build/api/types.go]: I thought the dockerconfig format was already capable of containing credentials for multiple repos

Comment (bparees) [inline pkg/build/api/types.go]: hm.  that is true.  though that doesn't necessarily mean you want all your creds in a single secret, but maybe it does make it sufficient.

Comment (liggitt) [inline pkg/build/api/types.go]: I guess the autopopulation aspect gets difficult... build admission has to find a secret that satisfies all the image pull specs (which there are now two of?)

Comment (bparees) [inline pkg/build/api/types.go]: yeah that would be another challenge w/ using a single field.

i'm definitely unsure about this since it's relatively unlikely people actually use two different secrets for this..but i could see someone saying "we want to use the openshift EAP image as the runtime image, but we have our own custom maven builder image that's on a separate registry" at which point you are in this scenario.

**Comment by PI-Victor** [comment]
@bparees when the s2i PR si merged, do we need a separate PR to bump Godeps, or can i make it a part of this but a separate commit?

File: pkg/build/api/v1/types.go (hunk 9 above)
Comment (PI-Victor) [inline pkg/build/api/v1/types.go]: @php-coder can you mention me on your pr when this gets added?

**Comment by bparees** [comment]
you can do it as a separate commit in this PR.  probably will make your life easier.

**Comment by PI-Victor** [comment]
@bparees thanks, that's why i was asking

File: pkg/build/api/v1/types.go (hunk 9 above)
Comment (php-coder) [inline pkg/build/api/v1/types.go]: @PI-Victor Ok, I'll mention you yesterday.

File: pkg/build/builder/sti.go [hunk 11]
```diff
@@ -384,3 +398,15 @@ func scriptProxyConfig(build *api.Build) (*s2iapi.ProxyConfig, error) {
 	}
 	return config, nil
 }
+
+// copyArtifactSourceList copies the artifacts set in the build config to the
+// VolumeList struct in the s2iapi.Config
+func copyArtifactSourceList(artifactsMapping []api.ImageSourcePath) (volumeList []s2iapi.VolumeSpec) {
```
Comment (php-coder) [inline pkg/build/builder/sti.go]: I suggest to change the type of return value from `[]s2iapi.VolumeSpec` to `s2iapi.VolumeList{}`. This tiny change will respect our abstraction.

Inspired by this code: https://github.com/openshift/origin/blob/2a2a8b9dfdc437c5b210fd7b748294ff4afe87ff/pkg/build/builder/sti.go#L148-L156

Comment (PI-Victor) [inline pkg/build/builder/sti.go]: good point, will do

File: pkg/build/builder/sti.go [hunk 12]
```diff
@@ -188,6 +198,9 @@ func (s *S2IBuilder) Build() error {
 		Injections:                injections,
 		ScriptDownloadProxyConfig: scriptDownloadProxyConfig,
 		BlockOnBuild:              true,
+		RuntimeImage:              runtimeImageName,
+		RuntimeArtifacts:          runtimeArtifacts,
+		RuntimePullSecret:         runtimePullSecret,
```
Comment (php-coder) [inline pkg/build/builder/sti.go]: I've added a field `RuntimeAuthentication`. As far I understand you shouldn't pass a secret to S2I directly, instead you're passing auth info. I've found this example and probably you could do the same: https://github.com/openshift/origin/blob/2a2a8b9dfdc437c5b210fd7b748294ff4afe87ff/pkg/build/builder/sti.go#L227-L228

File: pkg/build/api/validation/validation.go [hunk 13]
```diff
@@ -647,3 +649,20 @@ func validatePostCommit(spec buildapi.BuildPostCommitSpec, fldPath *field.Path)
...
+	}
+
+	if sourceStrategy.RuntimeImage.Name == "" {
+		allErrs = append(allErrs, field.Invalid(fldPath, sourceStrategy.RuntimeImage, "the runtime image name cannot be empty"))
+		return
+	}
+
+	if sourceStrategy.RuntimeImage.Name != "" && len(sourceStrategy.RuntimeArtifacts) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, sourceStrategy.RuntimeArtifacts, "at least one runtime artifacts mapping must be specified"))
+		return
```
Comment (php-coder) [inline pkg/build/api/validation/validation.go]: Just realized that we probably shouldn't do this check: when `RuntimeArtifacts` is empty, S2I will try to extract this mapping from the `io.openshift.s2i.assemble-input-files` label on a runtime image.

Comment (php-coder) [inline pkg/build/api/validation/validation.go]: Yes, I already have code that will error this out.

Comment (PI-Victor) [inline pkg/build/api/validation/validation.go]: ok, i'll remove then

File: pkg/build/builder/sti.go (hunk 8 above)
Comment (PI-Victor) [inline pkg/build/builder/sti.go]: is there any particular reason why, for example the From ref in sourcebuildstrategy is not a pointer yet DockerBuildStrategy is https://github.com/openshift/origin/blob/master/pkg/build/api/types.go#L475 and then, again Custom isn't  https://github.com/openshift/origin/blob/master/pkg/build/api/types.go#L444
are we concerned with mutating it?

should i not use a pointer after all for runtimeImage? https://github.com/openshift/origin/pull/9596/files#diff-f23c2970936ec5cfcb60d485ae96bcffR519

Comment (PI-Victor) [inline pkg/build/builder/sti.go]: @bparees ^^

Comment (bparees) [inline pkg/build/builder/sti.go]: @PI-Victor 
the From reference is required in a SourceBuildStrategy and Custom strategy, it is optional in a Docker build strategy (you can just have a Dockerfile that contains a FROM line).

we use pointers when things are optional so the value can be left nil.

**Comment by PI-Victor** [comment]
new PR here: https://github.com/openshift/origin/pull/9813
have to close this one since i deleted the old fork, long story.
